pkg/power/redfish/redfish.go

```diff
@@ -0,0 +1,255 @@
+	LastPowerOutputWatts int           `json:"LastPowerOutputWatts,omitempty"`
+	Model                string        `json:"Model,omitempty"`
+	Manufacturer         string        `json:"Manufacturer,omitempty"`
+	FirmwareVersion      string        `json:"FirmwareVersion,omitempty"`
+	SerialNumber         string        `json:"SerialNumber,omitempty"`
+	PartNumber           string        `json:"PartNumber,omitempty"`
+	SparePartNumber      string        `json:"SparePartNumber,omitempty"`
+	InputRanges          []InputRanges `json:"InputRanges,omitempty"`
+	RelatedItem          []RelatedItem `json:"RelatedItem,omitempty"`
+}
+type PowerPowerSupplyReset struct {
+	Target string `json:"target,omitempty"`
+}
+type Actions struct {
+	PowerPowerSupplyReset PowerPowerSupplyReset `json:"#Power.PowerSupplyReset,omitempty"`
+}
+
+// RedfishSystemPowerResult is the system power query result
+type RedfishSystemPowerResult struct {
+	system        string
+	consumedWatts int
+	timestamp     time.Time
+}
+
+// RedfishAccessInfo is the struct for the access model
+type RedfishAccessInfo struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+	Host     string `json:"host"`
+}
+
+type RedFishClient struct {
+	// systemEnergy is the system accumulated energy consumption in Joule
+	accessInfo RedfishAccessInfo
+	systems    []*RedfishSystemPowerResult
+	ticker     *time.Ticker
+	mutex      sync.Mutex
+}
+
+func NewRedfishClient(username, password, host string) *RedFishClient {
+	redfish := &RedFishClient{
+		accessInfo: RedfishAccessInfo{
+			Username: username,
+			Password: password,
+			Host:     host,
+		},
+	}
+	if redfish.IsPowerSupported() {
+		klog.Infof("starting redfish power collection\n")
+	}
+	return redfish
+}
+
+func (rf *RedFishClient) IsPowerSupported() bool {
+	system, err := getRedfishSystem(rf.accessInfo)
+
+	if err != nil {
+		klog.V(5).Infof("failed to get system info: %v\n", err)
+		return false
+	}
+
+	intervalInMin := 0
+	// iterate each "Members" in the system and get the power info
+	for _, member := range system.Members {
+		// split the OdataID by delimiter "/" and get the system ID
+		split := strings.Split(member.OdataID, "/")
+		if len(split) < 2 {
+			continue
+		}
+		id := split[len(split)-1]
+		res := RedfishSystemPowerResult{}
+		power, err := getRedfishPower(rf.accessInfo, id)
+		if err == nil && len(power.PowerControl) > 0 {
+			res.system = id
+			res.consumedWatts = power.PowerControl[0].PowerConsumedWatts
+			res.timestamp = time.Now()
+			rf.systems = append(rf.systems, &res)
+			klog.V(5).Infof("power info: %+v\n", power)
+			if power.PowerControl[0].PowerMetrics.IntervalInMin > intervalInMin {
+				intervalInMin = power.PowerControl[0].PowerMetrics.IntervalInMin
+			}
+		} else {
+			klog.V(5).Infof("failed to get power info: %v\n", err)
+		}
+	}
+
+	// set a timer to check the power info every intervalInMin minutes
+	if intervalInMin > 0 {
```

Comment:
`IntervalInMin` in Redfish is irrelevant to the interval of updating the value of `PowerConsumedWatts`.
I believe that `PowerConsumedWatts` is updated every one or few seconds. So it would rather be collected more frequently than using `IntervalInMin`.

Reply:
sounds good